Polymorphic types.

dkorolev opened the pull request:
Hi Max!

For after https://github.com/c5t/Current/pull/273.

Polymorphic types, all but serialization by now.

**Suggestion**: Looks like with a bit of tweaking, the dream may come true. Specifically:

* Use global `Exists(...)` or `Exists<T>(...)` for base types, optionals, and polymorphics.
* Use global `Value(...)` or `Value<T>(...)` for base types, optionals, and polymorphics.

We'll have to be careful with type hierarchies of types within `Polymorphic<>`. As of now, `Match()` would do the right thing with RTTI dispatching, while `Value<T>` would happily case derived to base if requested. The above does sound right to me, but want to double-check with you.

Thanks,
Dima

PS: Better naming suggestions for `Match()`? I don't like it. `MatchWith()` is more functional, but still doesn't sound right in C++.

mzhurovich commented on TypeSystem/Serialization/json.h:
```diff
@@ -212,160 +231,185 @@ struct LoadFromJSONImpl {
-    if (!source.IsArray()) {
+  static void Load(rapidjson::Value* source, std::vector<T>& destination, const std::string& path) {
+    if (source && source->IsArray()) {
+      const size_t n = source->Size();
```
`n`->`size`?

dkorolev replied:
:+1: 90329ba

mzhurovich commented on TypeSystem/polymorphic.h:
```diff
@@ -0,0 +1,182 @@
+
+  // Initializes an `std::unique_ptr<CurrentSuper>` given the object of the right type.
+  // The input object could be an object itself (in which case it's copied),
+  // an `std::move()`-d `std::unique_ptr` to that object (in which cast it's moved),
```
s/cast/case/

dkorolev replied:
:+1: 90329ba

mzhurovich commented on TypeSystem/polymorphic.h:
```diff
@@ -0,0 +1,182 @@
+  // an `std::move()`-d `std::unique_ptr` to that object (in which cast it's moved),
+  // or a bare pointer (in which case it's captured).
+  // TODO(dkorolev): The bare pointer one is sure unsafe -- consult with @mzhurovich.
+  struct TypeCheckedAsisgnment {
```
s/Asisgnment/Assignment/

dkorolev replied:
:+1: 90329ba

mzhurovich commented:
@dkorolev As for `Match()` - how about `Apply()` or `Call()` or `RTTICall()`?

dkorolev commented:
Renamed `Match()` into `Call()`, and removed that confusing `const` in the constructor parameter to `Optional<>`.

@mzhurovich PTAL.

Thanks,
Dima

mzhurovich commented on TypeSystem/polymorphic.h:
```diff
@@ -0,0 +1,184 @@
+
+  PolymorphicImpl() = delete;
+
+  // PolymorphicImpl(std::unique_ptr<CurrentSuper>&& rhs) : object_(std::move(rhs)) {}
```
Is it deliberately commented out?

dkorolev replied:
Fixed in 33fc228 !

mzhurovich commented:
@dkorolev Помолясь применимъ.

dkorolev commented:
Воистинъ Аминъ.
